### PR TITLE
Llt 5588 improve batching test

### PR DESCRIPTION
**Pull request by LukasPukenis**
### Problem
Improvements for the batching test. Commit contains more info

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

File: nat-lab/tests/test_batching.py [hunk 1]
```diff
@@ -43,82 +41,142 @@ def _generate_setup_parameters(
...
+# for an extended period. This idle period allows for a visual observation.
+# Local batching will only have an effect of batching multiple local keepalives into one bundle but will
+# not do anything about syncing the keepalives between the peers.
+
+
+# TODO: Add asserts for local batching
+# TODO: Implement received-data-trigger batching
 @pytest.mark.asyncio
-@pytest.mark.timeout(200)
+@pytest.mark.timeout(600)
```
Comment (gytsto) [inline nat-lab/tests/test_batching.py:128]: move value to `timeouts.py`

Comment (LukasPukenis) [reply nat-lab/tests/test_batching.py]: Done

File: nat-lab/tests/utils/batching.py [hunk 2]
```diff
@@ -84,20 +86,13 @@ def save_histogram(name: str, data: List[int], save_dir: str, max_height=None):
...
 
-    # try to save `name.hist` and warn if it already exists by asserting false
-    file_name = f"{name}.hist"
-    file_path = os.path.join(save_dir, file_name)
-
-    with open(file_path, "w", encoding="utf-8") as f:
-        f.write("\n".join(output))
-
-    print(f"Histogram saved successfully as '{file_name}' in '{save_dir}'.")
+    print("\n".join(output))
```
Comment (tomasz-grz) [inline nat-lab/tests/utils/batching.py:89]: Love some ascii art (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧

File: .github/workflows/gitlab.yml [hunk 3]
```diff
@@ -31,4 +31,4 @@ jobs:
           project-id: ${{ secrets.PROJECT_ID }}
           schedule: ${{ github.event_name == 'schedule' }}
           cancel-outdated-pipelines: ${{ github.ref_name != 'main' }}
-          triggered-ref: v2.4.0 # REMEMBER to also update in .gitlab-ci.yml
+          triggered-ref: lukasp/batching-tests-long # REMEMBER to also update in .gitlab-ci.yml
```
Comment (tomasz-grz) [inline .github/workflows/gitlab.yml:34]: Is this just for testing now?

File: nat-lab/tests/test_batching.py [hunk 4]
```diff
@@ -43,82 +42,142 @@ def _generate_setup_parameters(
...
+# This test captures histograms of network activity to evaluate the effect of local batching in libtelio.
+# Since only local batching is implemented, no client-generated traffic should occur during the test.
+# External traffic (incoming data) could distort the histograms, and receive-data-triggered batching is
+# not yet supported in libtelio. The test setup is simple: all clients are interconnected and remain idle
+# for an extended period. This idle period allows for a visual observation.
+# Local batching will only have an effect of batching multiple local keepalives into one bundle but will
+# not do anything about syncing the keepalives between the peers.
+
+
+# TODO: Add asserts for local batching
```
Comment (tomasz-grz) [inline nat-lab/tests/test_batching.py:126]: Will these come in a future PR?

File: nat-lab/tests/test_batching.py [hunk 5]
```diff
@@ -43,82 +42,142 @@ def _generate_setup_parameters(
...
-                    AdapterType.LinuxNativeWg,
-                    ["stun", "upnp", "local"],
-                    direct_enabled=True,
-                    batching_enabled=True,
-                ),
+                _generate_setup_parameters(conn_tag, adapter, False)
+                for conn_tag, adapter in ALL_NODES
+            ],
+            (0, 5),
+            240,
```
Comment (tomasz-grz) [inline nat-lab/tests/test_batching.py:139]: What do you think about declaring these as constant so it wold be easier from a glance to understand what the numbers mean?

File: nat-lab/tests/test_batching.py [hunk 6]
```diff
@@ -129,61 +188,63 @@ async def test_batching_star(
...
                 filter_name = filt[0] if filt else "none"
                 hs = generate_histogram_from_pcap(
                     pcap_path, capture_duration, filt[1] if filt else None
                 )
                 title = f"{container}-filter({filter_name})"
-                save_histogram(title, hs, "logs", max_height=12)
-
-        for p in pcap_paths:
-            os.unlink(p)
+                print_histogram(title, hs, max_height=12)
```
Comment (tomasz-grz) [inline nat-lab/tests/test_batching.py:250]: What are the test conditions here? or is this only intended for manual human inspection?
And do you think about having some rudimentary check like that histogram is not empty?

**Review by tomasz-grz — COMMENTED** [review]
Nice work! Just a few questions from my side

File: nat-lab/tests/utils/batching.py (hunk 2 above)
Comment (LukasPukenis) [reply nat-lab/tests/utils/batching.py:89]: There are dozens of us! :D

File: .github/workflows/gitlab.yml (hunk 3 above)
Comment (LukasPukenis) [reply .github/workflows/gitlab.yml:34]: I wanted to prove that it actually does something, so created accompanying PR in libtelio-build - https://bucket.digitalarsenal.net/low-level-hacks/vpn/client/libtelio-build/-/merge_requests/265 which filters for `batching` marked tests.

The regular `natlab` jobs wouldn't take this job.

Ideally I might revert it back before merging but probably no harm in even leaving it :)

File: nat-lab/tests/test_batching.py (hunk 6 above)
Comment (LukasPukenis) [reply nat-lab/tests/test_batching.py:248]: For now the plan is to just observe it and later on bolt on some assertions. This means manual inspection after nightly and/or manual runs :) 

This is not ideal and we could already do some asserts at this point but the CI runs are pretty slow so at this point I am satisfied by just having histograms somewhere for observation :)

File: nat-lab/tests/test_batching.py (hunk 5 above)
Comment (LukasPukenis) [reply nat-lab/tests/test_batching.py]: That's a good idea, I've extracted those to constants :) :+1:

Comment (tomasz-grz) [reply nat-lab/tests/test_batching.py]: Thanks

**Review by tomasz-grz — APPROVED** [review]
Looks good to me +1 🚀

File: .github/workflows/gitlab.yml (hunk 3 above)
Comment (LukasPukenis) [reply .github/workflows/gitlab.yml]: I've reverted those back

**Comment by Hasan6979** [comment]
+1 lgtm

**Review by Hasan6979 — APPROVED** [review]

**Review by tomasz-grz — APPROVED** [review]

**Review by Hasan6979 — APPROVED** [review]

**Review by jjanowsk — APPROVED** [review]

**Pull request merged**